### ISSUE TITLE
Add check that model overfits when training on a tiny dataset into ModelTestCase

### ISSUE THREAD

**Issue by mojesty** [allenai/allennlp#3394]
**Is your feature request related to a problem? Please describe.**
I'm writing tests for my own Models and every time before starting training on a large dataset I have to manually launch tests and ensure that model is able to overfit on my tiny testing dataset (I use same data for training and validation in my tests). I'm checking overfit by comparing `validation_metric` with some gold value (usually 1) with some tolerance and I think that this check is both helpful and simple to implement.

**Describe the solution you'd like**
Using `get_metrics` function every epoch to check that the given key (usually validation_metric or some default like accuracy) is close enough to desired value. If training stops (e. g., we ran out of epochs/patience) and metric is not close enough, raise error to show that something is not right.

**Additional context**
Although this check does not guarantee that model will successfully train on a large dataset, usually it's a clear indicator of some problem in the model. To the best of my knowledge, simple 32-cell lstms have enough capacity to overfit to 2-10 Instances datasets and are relatively fast to train, so I do not expect big computational overhead after implementing this.

I am interested in implementing this and can write this myself if you are interested in it too.

**Comment by brendan-ai2** [comment]
This sounds like great idea, @mojesty. We'd be happy to take such a contribution. Could you add this feature to https://github.com/allenai/allennlp/blob/master/allennlp/common/testing/model_test_case.py#L15? Thanks!

**Comment by mojesty** [comment]
I started to implement this and have several questions regarding this check:

- sometimes there is not `validation_metric` key in `Trainer` init arguments. In this case, `loss` is used, so I suggest in this case checking that loss is close to 0 does the job
- sometimes validation metrix is provided, but it is not obvious which value corresponds to overfitting. For simple accuracy / f measure, that is obviously 1, but for other metrics (e. g. BLEU) the overfit-indicating value is greater than 1.

So it seems to me that the only simple way is to check loss value on validation dataset that is saved in `metrics.json` after model training. Other variants like checking accuracy requires additional specification that the model creator may not want to specify in test suite.

**Comment by brendan-ai2** [comment]
Hi again @mojesty, good point about different metrics and different overfit values. (Perhaps this is a dumb question, but is it really true in general that an overfit model will have its loss driven to zero?) To get around this how about the test function simply takes a `metric` to optimize and a `terminal_value` to be driven towards. This would still be pretty easy for model creators to use while being much more robust.
